WordPress/Classes/System/Root View/WordPressAuthenticatorProtocol.swift

```diff
@@ -8,20 +9,57 @@ protocol WordPressAuthenticatorProtocol {
 
 extension WordPressAuthenticator: WordPressAuthenticatorProtocol {
     static func loginUI() -> UIViewController? {
-        Self.loginUI(showCancel: false, restrictToWPCom: false, onLoginButtonTapped: nil, continueWithDotCom: { viewController in
-            guard Self.dotComWebLoginEnabled, let navigationController = viewController.navigationController else {
-                return false
-            }
+        Self.loginUI(
+            showCancel: false,
+            restrictToWPCom: false,
+            onLoginButtonTapped: nil,
+            continueWithDotCom: { viewController in
+                guard Self.dotComWebLoginEnabled, let navigationController = viewController.navigationController else {
+                    return false
+                }
 
-            Task { @MainActor in
-                let accountID = await WordPressDotComAuthenticator().signIn(from: navigationController, context: .default)
-                if accountID != nil {
-                    WordPressAppDelegate.shared?.presentDefaultAccountPrimarySite(from: navigationController)
+                Task { @MainActor in
+                    let accountID = await WordPressDotComAuthenticator().signIn(from: navigationController, context: .default)
+                    if accountID != nil {
+                        WordPressAppDelegate.shared?.presentDefaultAccountPrimarySite(from: navigationController)
+                    }
                 }
-            }
 
-            return true
-        })
+                return true
+            },
+            selfHostedSiteLogin: { viewController in
+                guard FeatureFlag.authenticateUsingApplicationPassword.enabled else { return false }
+                guard let window = viewController.view.window, let navigationController = viewController.navigationController else { return false }
+
+                let client = SelfHostedSiteAuthenticator(session: URLSession(configuration: .ephemeral))
```

Comment:
Please extract it into a separate method. The level of nesting is unwieldy.

(nit) I'd also refactor this a bit into something like:

```
onContinueTapped: (selection: Selection) -> Void

enum Selection {
    case dotCom
    case selfHosted
}
```